pkg/kubelet/cri/remote/remote_runtime_test.go

```diff
@@ -68,6 +68,8 @@ func TestVersion(t *testing.T) {
 	r := createRemoteRuntimeService(endpoint, t)
 	version, err := r.Version(apitest.FakeVersion)
 	assert.NoError(t, err)
-	assert.Equal(t, apitest.FakeVersion, version.Version)
-	assert.Equal(t, apitest.FakeRuntimeName, version.RuntimeName)
+	if version != nil {
+		assert.Equal(t, apitest.FakeVersion, version.Version)
+		assert.Equal(t, apitest.FakeRuntimeName, version.RuntimeName)
+	}
```

Comment:
```suggestion
	require.NoError(t, err)
	require.NotNil(t, version)
	require.Equal(t, apitest.FakeVersion, version.Version)
	require.Equal(t, apitest.FakeRuntimeName, version.RuntimeName)
```

require will end the test early rather than stacking failures, which I think acomplishes the same thing you were going for here? but without the potential for accidental false positives (if err && version == nil)

Reply:
> accidental false positives

Good question.. I'm not against this test failing on nil in certain circumstances.. but this is tricky and I don't think the bucket is ready/well written for all the possible expected cases. Let me explain.

In the "field" the timing of (1) start runtime, (2)- connect to grpc socket remote, and (3) start making calls to the remote.. requires some delay before r.Version() and other apis are expected to actually work correctly. This is because of grpc timeouts and the runtime posting up the available grpc services in it's init cycle before it is finished with all of it's init work and actually ready for grpc server requests. There's a timing window here to be resolved. Typically we add a until "remoteService.Version() == nil; sleep(1sec)" type loop on clients to make sure we are actually talking to the server.

Clearly the intent of this test is to check if the fake runtime returns the expected meta.. and depending on system timing issues you might get a grpc timeout or bad connection instead and that gets swallowed.  We could add a sleep loop and wait on every cri api to return non nil. We could address this by considering if we need to add a client side error when grpc errors occur. We could also modify NewFakeRemoteRuntime or createRemoteRuntimeService to wait some time until the runtime is actually ready. 

IOW a slow startup or other grpc connection issue is not necessarily an error, unless it has a sleep loop in the test and the grpc induced timeouts go beyond some reasonable amount of time. The ability of the client to be considerate of a not-actually ready runtime or temporary grpc disconnect is an issue that needs some number of buckets for testing understanding and resolution. 

Another issue, should not avail (nil) be supported at all? The API does not specify...

I could add a "TODO: (Mike Brown) refactor for startup timing issues" if you like

Reply:
TODO sounds good @mikebrow

Reply:
SGTM